### DISCUSSION TITLE
Unit Testing

### DISCUSSION THREAD

**Discussion by roboteng** [Rahix/avr-hal#379] [Ideas]
I'm not sure where to post this, but I think it could be helpful here.  I am new to rust, so I am not 100% proficient with it, but I have not found a way to do unit testing.  I'd totally be willing to write up some documentation once I get it working. I haven't found any references to testing within this repo, nor within the wider rust community.

Any help would be appreciated, and I'd love to make a PR for this repo or the template one if I can figure it out.

**Comment by Patryk27** [comment]
I can do a bit of self-advertisement here:
https://github.com/Patryk27/avr-tester/ 🙂 

It's not as straightforward as `cargo test`, but it's very close!

**Comment by djdisodo** [comment]
you can use qemu, it works
https://github.com/Rahix/avr-hal/discussions/280
it works without no real hardware, yet will have timer issue
https://github.com/Rahix/avr-hal/pull/257
